### DISCUSSION TITLE
ECC support (detection & polling)

### DISCUSSION THREAD

**Discussion by 01e3** [memtest86plus/memtest86plus#92] [Enhancement]
Memtest86+ needs to handle ECC / EDAC to properly support testing ECC memory.

Otherwise, bad memory may end up being reported as good due to errors being masked by the ECC mechanism.

**Comment by oed-streicherd** [comment]
Yes - EDAC, ECC-Event reporting support would be nice, on server systems with activated ECC we have to check the manufacture system event logs to see if ECC errors happen.
~ This is okay, as far as you know that you have to look at system logs. 
The memtest screen keeps "clean", because events are masked by ECC/BIOS mechanism.

**Comment by x86fr** [comment]
ECC/EDAC support requires specific coding for each memory controller, so this is too generic for an issue. 

Enhancement ideas without an active developer working on it should be posted in discussions, so the users can upvote them and tell us why they need it.

That said, ECC support is important, but I think this will not be implemented before v7.

**Comment by Fantu** [comment]
it would be useful on servers (mostly with ddr memories) and some workstations (quite rare), however I understand it would take a lot of work to implement. if the errors are not detected by memtest86+ because corrected by the ECC system then even if it is useful to detect if there are error correct by ECC, the current identification of errors not even corrected by ECC is still sufficient (uncorrectable errors are those that cause significant problems) and therefore I think that this taks is not a priority (FWIK)

**Comment by Craig-Macomber** [comment]
All Ryzen 7000 series CPU and many full systems have official ECC support and as of AGESA 1.0.0.5 patch C, it actually works (That was 2023/02/24 for my motherboard). Intel Alder Lake also has ECC support on many CPU (both desktop and mobile) with the W680 which launched in Q1'22.

Thus I think ECC is becoming more relevant.

Personally I want to use ECC error detection + some memory tester to help detect the threshold which I can safely overclock my ram (ECC memory has slow default speeds, but at least mine happily runs at more more gamer marked ram speeds).

Fortunately correctable memory errors aren't destructive to a an OS, so using a user-space memory tester (or just regular use) and OS provided ECC error reporting is a solid alternative in this use-case. Thus I could deciding ECC support (which looks like a ton of work) is out of scope for MemTest+.

**Comment by x86fr** [comment]
I've started looking at it actively. Anyway, don't expect a first ECC support before v7. I still have a lot of work on #280 before really starting to work on ECC code.

The first target platforms will likely be Zen 3/4 and ADL/SPR depending on which of AMD or Intel will help first. Both offered to help on ECC support by providing hardware tools that will make the debugging process much easier.

**Comment by aidan-gibson** [comment]
Thanks for your work on this! What is the current status?

**Reply by edo1** [reply]
@aidan-gibson 
It should work for some AMD CPUs. But I haven't been able to get it to work yet.
https://github.com/memtest86plus/memtest86plus/discussions/436

**Comment by JoeSchmuck** [comment]
It has been a while since someone has asked about ECC Support and the documentation really does not list anything with respect to ECC Support.  [The last thing I read had to do with version 7 and where a person had to rebuild the executable with a value set to true. ](https://github.com/memtest86plus/memtest86plus/discussions/436)

Is this the same with version 8.10 or is there a switch or other way to enable this feature, or is it automatically enabled by default?

And if a person still must compile the program separately, can a binary be offered up for people to download at memtets86.org ?

**Reply by x86fr** [reply]
With v8.10+, you can enable ECC polling at startup. F1 to display the configuration menu, then F8 (only available at startup!), then F2 to enable ECC.

**Comment by JoeSchmuck** [comment]
Thank you for that comment.  I will give that a try in the morning.

**Comment by edo1** [comment]
> F1 to display the configuration menu, then F8 (only available at startup!)

It looks like F8 is bound to a different action here

<img width="2560" height="1166" alt="image" src="https://github.com/user-attachments/assets/c3ee69ca-6a00-41d4-af7a-a46dd1aaf398" />

**Comment by x86fr** [comment]
Yes. ECC polling option on startup menu has been added with f9a30b0fa6737145dfd72d9951a62eeafb25e512 and is currently only available on nightly builds (until v9 release)

**Reply by JoeSchmuck** [reply]
Thank you for clarifying, I would have understood your previous statement to mean version 8.10 and up.  I grabbed the 07/12/2026 nightly and will give it a try.

**Reply by x86fr** [reply]
Let me know if you find an issue. Your Ryzen 7 7700 (Raphael) is supported.

**Comment by JoeSchmuck** [comment]
Okay, I have a stupid question:  How can I tell that ECC is actually being detected and tested?  I don't see any difference in the nightly build (I just downloaded last nights version).

I did see the F8 Boot menu to F2 ECC "enable/disable".  It appears to default to "enable".  I toggled it and left it on "eneblee", yes, there is a minor grammar error in the software.  Looks like the leftover "e" is from "disable" and the last "e" remains.

But, I cannot tell that ECC is being tested.  Maybe there is no indication set in the current build, but planned for later implementation.  The SPD clearly indicates the RAM is ECC, and it does in 8.x as well, and I think 7.x.  It is just what the SPD says.  This does not indicate ECC is active or being tested.

Otherwise, it seems to run fine on my test computer.  I only have 16GB ECC RAM in it.

**Reply by x86fr** [reply]
If ECC is enabled, you will see two Error counters on main screen : `Err:` and `ECC:`
Please post a screenshot if any doubt.

**Comment by JoeSchmuck** [comment]
I'm probably blind?  I did select F1, then F8, and I can see ECC was enabled.  For the heck of it I disabled it and the screen still does not indicate ECC being tested.  I am running a A1SAM-2550F with unbuffered ECC RAM.  The board automatically detects ECC RAM and utilizes it, no BIOS setting.

<img width="2792" height="1601" alt="No-ECC" src="https://github.com/user-attachments/assets/ddf9588d-a8a0-413d-9381-62237e4fa63a" />

**Reply by x86fr** [reply]
This old Atom is not supported. ECC checking requires specific proprietary code for every different CPUs and is currently only supported on some CPUs (eg: most AMD Zen+ CPUs, like the Ryzen 7 7700 you posted previously).

**Reply by JoeSchmuck** [reply]
I did not post my CPU, that was someone else, otherwise I would have stated the Intel Atom CPU.  This is my test machine where I can corrupt any or all the data, destructive tests, etc.

So only AMD?  That sounds pretty limitating to many of us who love this program.  I did verify my system has ECC enabled while running Debian using the following command `sudo dmidecode -t memory | grep -E "Total Width|Data Width"`

Of course being able to test if ECC is enabled vice being able to test it are two different things.

Are there any plans to support Intel?  My current home server is an AMD Ryzen 5500G, I will test this program out on it, but my "test machine" which is the Intel Atom, will not work unfortunately.

And thank you for all the effort maintaining this wonderful program.  I recommend it all the time, but now when it comes to ECC, which most people are using in the TrueNAS world, they may not be able to test their ECC RAM, and a lot of these folks use Intel.

Cheers
